### ISSUE TITLE
Using local browser for generation consistently runs at ~1 IT /s lower speed than LAN remote from phone browser, both slower than CLI

### ISSUE THREAD

**Issue by NeedsMoar** [nod-ai/SHARK#1457]
Summary:   Web server appears to be eating up far more time than it should and hurting performance when used locally by a measurable amount. (I suspect some kind of polling speed on the page that's hammering the server when not network speed  limited?).  Both are slower than the CLI as expected, but the remote browser is closer to the CLI in speed.

Version:    I'm still using an old version, _20230329_648_ but had this issue in every version I tried previously and also tested 700 for this bug where I found another weird issue.

CLI being faster is expected, no probably horrifying mess of python of a web server to run in the background and constantly switch context to.  Shouldn't matter if the multithreading is done right but it never is, in some library or another.  Maybe it's just the unlimited bandwidth and a javascript without a timer.  I don't particularly want to dig into it.  :D

With the same prompts, local firefox connection to the site with only the site tab open, on a tuned stablediffusion-2.1-base:
These are 9-image batches, I only posted the last few and total time of each that listed it for brevity's sake:

**648**
Local:
```
Clip Inference time (ms) = 32.007
Average step time: 46.71104431152344ms/it
VAE Inference time (ms): 161.386
Clip Inference time (ms) = 35.009
Average step time: 46.708197593688965ms/it
VAE Inference time (ms): 161.305
Clip Inference time (ms) = 35.008
Average step time: 46.80663585662842ms/it
VAE Inference time (ms): 155.993
Total image(s) generation time: 26.0507sec
```

Running in android browser:
```
Clip Inference time (ms) = 44.631
Average step time: 42.3353385925293ms/it
VAE Inference time (ms): 150.511
Clip Inference time (ms) = 29.171
Average step time: 42.669034004211426ms/it
VAE Inference time (ms): 152.209
Clip Inference time (ms) = 29.366
Average step time: 42.41556167602539ms/it
VAE Inference time (ms): 152.488
Total image(s) generation time: 24.0468sec
```

Nearly 8% slower local webpage is pretty bad.

Results are roughly similar for any given run.  

The CLI version produces the following for the same parameters, averaged a bit faster than the remote browser as might be expected with it not having to manage piping data over to a web server. I didn't post the total time because the CLI incorrectly showed it as the time of the last image instead of the batch on this version.
```
Clip Inference time (ms) = 43.335
Average step time: 43.91268730163574ms/it
VAE Inference time (ms): 152.439
Clip Inference time (ms) = 31.007
Average step time: 42.06586837768555ms/it
VAE Inference time (ms): 150.698
Clip Inference time (ms) = 35.008
Average step time: 42.136216163635254ms/it
VAE Inference time (ms): 147.709
```

I bit the bullet and installed the current release (20230423_700) but there doesn't seem to be a change except a very slight regression in speed for the GUI versions in general that I'd need to gather more data to verify. 

**700**

```
Clip Inference time (ms) = 90.571
Average step time: 46.40392303466797ms/it
VAE Inference time (ms): 151.710
Clip Inference time (ms) = 98.910
Average step time: 46.31359577178955ms/it
VAE Inference time (ms): 158.418
Clip Inference time (ms) = 92.021
Average step time: 46.401963233947754ms/it
VAE Inference time (ms): 155.235
Total image(s) generation time: 26.4638sec
```

Remote browser is also slightly slower across several runs.  
```
Clip Inference time (ms) = 111.076
Average step time: 42.43260383605957ms/it
VAE Inference time (ms): 149.256
Clip Inference time (ms) = 99.577
Average step time: 42.79223442077637ms/it
VAE Inference time (ms): 150.784
Clip Inference time (ms) = 108.680
Average step time: 43.15391540527344ms/it
VAE Inference time (ms): 150.699
Total image(s) generation time: 24.8097sec
```

6% faster remotely on this run...  

CLI:
```
Clip Inference time (ms) = 101.023
Average step time: 41.77597999572754ms/it
VAE Inference time (ms): 146.647
Clip Inference time (ms) = 98.022
Average step time: 42.27846145629883ms/it
VAE Inference time (ms): 149.794
Clip Inference time (ms) = 98.023
Average step time: 41.8491268157959ms/it
VAE Inference time (ms): 152.014
```

Still only shows generation time for single image, but the CLI speed seems to have slightly improved.  Not enough to justify the 33% power increase.  Sorry I'm only showing a small amount of data here, I don't feel like running this hundreds of times and graphing it right now but there's a definite trend.  

The main thing I'm noticing is that the 648 version would usually top out the clock speed at the max boost frequency (on my XFX board, it defaults it to 2985 with the "quiet" BIOS but I think that's a driver thing) and stay there if the load was heavier and constant (3rd party / non-tuned models) or rapidly jump for each batched image with the tuned models which it usually cranked out fast enough that it didn't stay high for long.  The memory speed stayed at about half the set max (1.3GHz in my case).   Board power hovered at 300W regardless.   The new version rapidly maxes out both memory and GPU clock, the board hits the hard limit of 425W in the non-OC bios, and GPU clock is dropped 100-200MHz to compensate, which ends up making everything slower. 

So aside from the bug, I guess my other question would be exactly what changed between 648 & 700 that caused the huge power draw increase without obvious speed increases.  More particularly why it's suddenly using memory at full blast when it was looking like a mostly compute load before.  Actual in-use memory on the GPU is exactly the same between versions.   If there's an improvement in txt2img that I haven't seen because I only tested for speed and power, that's great too.   _Just wondering here._ 

Hardware is 
- Threadripper PRO 5975WX, NPS1, NUMA as SRAT off, Virtualization Off
- 512GB (64GBx8) DDR4 3200
- 2 x 2TB PCIe 4.0 NVMes for OS / Programs / Video Editing, 1 x 2TB PCIe 3.0 NVMe, 1x 16TB HDD (drives aren't relevant, Windows memory caching has all the models stored after first compile and they're only loaded from disk again if I've edited so many camera raws between runs that the model fell out of the 512GB of ram, or ran some memory hog software.
- XFX Radeon 7900XTX (whatever stupid name they gave the non-reference version)
- Windows 10 Pro Workstation 
- Network, etc, probably unimportant, but right now it's a 1Gb -> wifi router -> phone for the devices discussed above.  Those aren't the issue.

**Comment by NeedsMoar** [comment]
I ran the firefox profiler because I hate myself (I'd rather fall on an iron fence from 30 feet up than look at the mess that is JavaScript) and generating 8 images has ~56% of time in page redraws from updates to output boxes and percentages with no apparent limit on quantity.  Something seems off there regardless of speed...   updating a line of text and appending to a text box shouldn't be eating up that time in redraws so I don't know if some weird styling is happening that's causing it to render in a more complex way or if firefox is just bad at this.  30% more is in FrameRequestCallback + promiseCallback + MutationCallback which is pretty nasty given how little this info is really needed.    I'm not sure that it's actually eating up that much time or if it's just the shader context switches to render text that are hurting things there. 

I didn't see any way of limiting refresh speed for those callbacks in GradIO when the textbox is being set up (there's a timer on the selection & copy function which is handled in JS for some reason, which also seems unnecessary;  plus my normal browser security setup disables the ability of javascript to access the clipboard so just using the native text selection and controls is more compatible)   but I'm not familiar with the codebase and don't want to be so somebody else might know a simple way.  

`DisplayList building 66%
WebRender display list 28%
Other 4.1%`

As far as I can tell it's faster over the network because the amount of data being funneled into the browser is throttled by the connection itself.  

Minimizing the browser Window during generation immediately jumps to CLI speeds.  I tend to suspect Node.js since it was written more to be easy to deal with than to do anything well, although it basically ruined Adobe Creative Cloud speeds when they included it as part of everything in addition to writing out 100s of megs of rolling logs for things like failed connections to corporate license management servers when a single user license is installed in the temp directory which was eroding the NVMe with the windows temp directory.

I know the models are monolithic so the lowering to GPU machine language probably can't be parallelized, but the 3-6 minutes doing the huge single-core lower with no feedback in the CLI could probably be improved, at least to show Clang / IREE / whatever output or a progress within each file.  I know it might not be doable but it's sad to see 30 CPU cores sitting idle while that long process happens any time size is changed followed by images being generated in seconds each.  Even my old machine was 10 core and the 16 core Ryzens seem to be the most popular among consumer CPUs.  Is it possible to lower all parts of the model at the same time?  (checkpoint, VAE, solver, etc)?  That's a lesser problem since it doesn't get changed often but the browser thing seems like low-hanging fruit given the 6-10% performance hits just from not minimizing the browser window.    :-)

**Comment by NeedsMoar** [comment]
723 with the UI went back to not maxing out VRAM.  GPU clock hit a record of just over 3GHz even though boost is set lower than that, so presumably performance is back to normal.  I'm messing with 640x768 images right now so they're massively slower and I haven't compared yet.

**Comment by powderluv** [comment]
Thank you for the detailed analysis. We plan to revisit tuning all different sizes for the 7900xtx soon so it should help with the non 512x512 cases.

**Comment by NeedsMoar** [comment]
Speeds are quite a bit faster for 512x768 now but I'm still getting the differences with the web UI minimized / remote vs. maximized, for example, web UI open:
50it [00:04, 12.28it/s]
50it [00:04, 12.21it/s]
50it [00:04, 12.25it/s]
50it [00:04, 12.20it/s]

web UI minimized or remote:
50it [00:03, 13.24it/s]
50it [00:03, 13.10it/s]
50it [00:03, 13.08it/s]
50it [00:03, 13.12it/s]

Compilation has gotten much faster over the course of my using this.  I raised the num_iters parameter from the default but I'm not sure how that affects things.  

I haven't tried 768x768 yet on .786;  last time I did there was no tuning for it and it produced what looked like scaled RGB noise at 1 IT/s.  The flatfile was 1GB, around half the size of the tuned ones.   640x640 produced a .vmfb that was <100MB vs the 2GB for sizes that had tuning info, and ran at 22s/it (and just produced a different, stretched version of the RGB noise.

**Comment by one-lithe-rune** [comment]
I think the minimized/remote vs maximized difference is something to do with the gradio progress animations.

If I run with `--no-progress_bar` then the first image of each batch takes the same (faster) time regardless of whether the browser is minimized or maximized.

Once the first image is in the batch is generated and put in the gallery, that gallery and the status box starting get the pulsing animation outline -- which I thought was one of the things `--no-progress_bar` disabled, maybe the gradio upgrade to 3.34.0 broke it? -- and further images get the slower time unless the windows in minimized.

**Comment by one-lithe-rune** [comment]
Adding the following CSS removes the pulsing animation, and makes everything run at the minimized speed for me.

```
.generating {
   animation: none !important;
}
```

**Comment by NeedsMoar** [comment]
This appears to be fixed by @one-lithe-rune 's checkin.  I'm closing the issue.

**Comment by NeedsMoar** [comment]
Fixed in #1613
